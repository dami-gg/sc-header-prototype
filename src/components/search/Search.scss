@import "../../../node_modules/dress-code/dist/sass/dress-code";

.search-form {

  .dc-search-form {
    margin-bottom: 0;
    width: 100%;

    &__input {
      box-sizing: border-box;
      color: $dc-black;
      width: 100%;
    }
  }

  .dc-suggest.visible {
    opacity: 1;
  }
}

