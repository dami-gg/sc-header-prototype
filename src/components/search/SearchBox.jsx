import React, {Component} from 'react';

import './Search.scss';

class SearchBox extends Component {
  render() {
    return (
        <input className="dc-input dc-search-form__input"
               onChange={(event) => {
                 return this.props.onChangeAction && this.props.onChangeAction(event)
               }}
               placeholder="Search..."
        />
    );
  }
}

SearchBox.propTypes = {
  onChangeAction: React.PropTypes.func
};

export default SearchBox;
