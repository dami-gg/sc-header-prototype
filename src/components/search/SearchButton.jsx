import React, {Component} from 'react';

import './Search.scss';

class SearchButton extends Component {
  render() {
    return (
        <button className="dc-btn dc-search-form__btn">
          <i className="dc-icon dc-icon--search dc-icon--interactive"></i>
        </button>
    );
  }
}

export default SearchButton;
