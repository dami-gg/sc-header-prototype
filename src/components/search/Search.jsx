import React, {Component} from 'react';
import {connect} from 'react-redux';

import './Search.scss';
import SearchBox from './SearchBox';
import SearchResults from './SearchResults';
import SearchButton from './SearchButton';
import {updateBusinessPartnerSearchResults} from '../../actions';

export class Search extends Component {
  componentWillMount() {
    this.props.updateBusinessPartnerSearchResults([]);
  }

  render() {
    return (
        <div className="dc-search-form">
          <SearchBox
              onChangeAction={this.updateResults.bind(this)}
          />
          <SearchButton />
          {
            this.props.searchResults && this.props.searchResults.length > 0 &&
            <SearchResults
                searchResults={this.props.searchResults}
                onClickAction={this.props.onClickAction}
            />
          }
        </div>
    );
  }

  updateResults(event) {
    let search = event.target.value;
    let regex = new RegExp(search, "gi");

    let searchResults = this.props.searchableItems.filter((item) => {
      return regex.test(item.name);
    });

    this.props.updateBusinessPartnerSearchResults(searchResults);
  }
}

Search.propTypes = {
  searchableItems: React.PropTypes.array.isRequired,
  onClickAction: React.PropTypes.func
};

const mapStateToProps = (state) => ({
  searchResults: state.businessPartners.businessPartnerSearchResults
});

const mapDispatchToProps = (dispatch) => {
  return {
    updateBusinessPartnerSearchResults: (businessPartnerSearchResults) => {
      dispatch(updateBusinessPartnerSearchResults(businessPartnerSearchResults));
    }
  }
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(Search);
