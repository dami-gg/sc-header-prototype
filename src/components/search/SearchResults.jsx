import React, {Component} from 'react';

import './Search.scss';

class SearchResults extends Component {
  render() {
    let resultNodes = this.props.results.map((result) => {
      return (
          <li
              className="dc-suggest__item dc-link"
              onClick={() => {
                return this.props.onClickAction && this.props.onClickAction(result);
              }}
              key={result.id}
          >
            {result.name}
          </li>
      );
    });

    return (
        <div>
          <ul className={'dc-list dc-suggest' + (this.props.results.length > 0 ? ' visible' : '')}>
            {resultNodes}
          </ul>
        </div>
    );
  }
}

export default SearchResults;
