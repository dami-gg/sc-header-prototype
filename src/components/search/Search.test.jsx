import React from 'react';
import {shallow} from 'enzyme';
import {shallowToJson} from 'enzyme-to-json';

import {Search} from './Search';
import SearchBox from './SearchBox';
import SearchResults from './SearchResults';
import SearchButton from './SearchButton';

const mockedSearchResults = [{
  id: 1,
  name: 'Result 1'
}, {
  id: 2,
  name: 'Result 2'
}];

function setup(hideSearchResults) {
  const props = {
    // State props
    searchResults: hideSearchResults ? [] : mockedSearchResults,

    // Dispatch props
    updateBusinessPartnerSearchResults: jest.fn()
  };

  const enzymeWrapper = shallow(
      <Search
          searchableItems={[]}
          {...props}
      />
  );

  return {
    props,
    enzymeWrapper
  }
}

describe('Search component', () => {
  it('should render correctly', () => {
    const { enzymeWrapper } = setup();

    expect(shallowToJson(enzymeWrapper)).toMatchSnapshot();
  });

  it('should render the SearchBox sub component', () => {
    const { enzymeWrapper } = setup();

    const searchBox = enzymeWrapper.find(SearchBox);
    expect(searchBox).not.toBeUndefined();

    const searchBoxProps = searchBox.props();
    expect(searchBoxProps.onChangeAction).toBeTruthy();
  });

  it('should render the SearchButton sub component', () => {
    const { enzymeWrapper } = setup();

    const searchButton = enzymeWrapper.find(SearchButton);
    expect(searchButton).not.toBeUndefined();
  });

  xit('should render the SearchResults sub component when there are search results to show', () => {
    const { enzymeWrapper } = setup();

    const searchResults = enzymeWrapper.find(SearchResults);
    expect(searchResults).not.toBeUndefined();

    const searchResultsProps = searchResults.props();
    expect(searchResultsProps.searchResults).toBeTruthy();
    expect(searchResultsProps.onClickAction).toBeTruthy();
  });

  xit('should not render the SearchResults sub component when there are no search results to show', () => {
    const { enzymeWrapper } = setup(true);

    const searchResults = enzymeWrapper.find(SearchResults);
    expect(searchResults).toBeUndefined();
  });

  it('should update the results after the SearchBox sub component triggers an onChangeAction event', () => {
    const { enzymeWrapper, props } = setup();
    const mockedEvent = {
      target: {
        value: 'Result 1'
      }
    };

    const searchBox = enzymeWrapper.find(SearchBox);
    searchBox.props().onChangeAction(mockedEvent);

    expect(props.updateBusinessPartnerSearchResults.mock.calls).toBeDefined();
  });
});