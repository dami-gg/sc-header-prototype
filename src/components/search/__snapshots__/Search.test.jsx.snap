exports[`Search component should render correctly 1`] = `
<div
  className="dc-search-form">
  <SearchBox
    onChangeAction={[Function bound updateResults]} />
  <SearchButton />
  <SearchResults
    onClickAction={undefined}
    searchResults={
      Array [
        Object {
          "id": 1,
          "name": "Result 1"
        },
        Object {
          "id": 2,
          "name": "Result 2"
        }
      ]
    } />
</div>
`;
